app/uk/gov/hmrc/gform/obligation/ObligationService.scala

```diff
@@ -17,30 +17,105 @@
+                              formTemplate,
+                              FormDataHelpers.formDataMap(form.formData),
+                              form.envelopeId))
+      replaceIdNumber = listOfIdNumbers.zip(hmrcTaxPeriodIdentifiers)
+      output <- gformConnector
+                 .getAllTaxPeriods(replaceIdNumber.map(a =>
+                   HmrcTaxPeriod(a._2.idType, TextExpression(Constant(a._1)), a._2.regimeType)))
+                 .map(i =>
+                   i.flatMap(j => j.obligation.obligations.flatMap(h => makeAllInfoList(j.id, h.obligationDetails))))
+    } yield output
+  }
+
+  def makeAllInfoList(id: HmrcTaxPeriod, obligation: List[ObligationDetail]) =
+    obligation.map(i =>
+      TaxPeriodInformation(id, i.inboundCorrespondenceFromDate, i.inboundCorrespondenceToDate, i.periodKey))
+
+  def lookupIfPossible(
+    form: Form,
+    formTemplate: FormTemplate,
+    authService: AuthService,
+    retrievals: MaterialisedRetrievals)(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
     val hmrcTaxPeriodIdentifiers = formTemplate.expandFormTemplateFull.allFCs.collect {
       case IsHmrcTaxPeriod(el) => el
     }
-    val futureListOfTaxPeriodDes =
-      gformConnector
-        .getAllTaxPeriods(hmrcTaxPeriodIdentifiers)
-        .map(i => i.flatMap(j => j.obligation.obligations.flatMap(h => makeMap(j.id, h.obligationDetails))))
-    futureListOfTaxPeriodDes.map(i => i.toMap)
+    for {
+      string <- authService.evaluateSubmissionReference(
+                 hmrcTaxPeriodIdentifiers.head.idNumber,
+                 retrievals,
+                 formTemplate,
+                 FormDataHelpers.formDataMap(form.formData),
+                 form.envelopeId)
+      output <- form.obligations match {
+                 case Some(x) => {
+                   for {
+                     idNumbers <- Future.traverse(x.listAllInfo)(
+                                   i =>
+                                     authService.evaluateSubmissionReference(
+                                       i.hmrcTaxPeriod.idNumber,
+                                       retrievals,
+                                       formTemplate,
+                                       FormDataHelpers.formDataMap(form.formData),
+                                       form.envelopeId))
+                     output <- {
+                       if (idNumbers.forall(i => string.equals(i))) {
+                         Future.successful(form)
+                       } else {
+                         val newObligations = lookupObligationsMultiple(formTemplate, authService, retrievals, form)
+                         newObligations.map(i => form.copy(obligations = Some(ListOfTaxPeriodInformation(i))))
+                       }
+                     }
+                   } yield output
+                 }
+                 case None => {
+                   if (hmrcTaxPeriodIdentifiers.forall(i => !i.regimeType.value.isEmpty && !string.isEmpty)) {
```

Comment:
This `if ... else` is that same (except condition) as the `if ... else` starting on line 92. Let's reuse the one code for both cases.